File: android/app/src/main/kotlin/com/torvm/android/packet/IPv4Header.kt
package com.torvm.android.packet

/**
 * Parses and builds IPv4 headers (minimum 20 bytes).
 *
 * All multi-byte fields use network byte order (big-endian).
 * IP addresses are stored as ByteArray(4).
 */
data class IPv4Header(
    val version: Int,
    val ihl: Int,
    val tos: Int,
    val totalLength: Int,
    val identification: Int,
    val flags: Int,
    val fragmentOffset: Int,
    val ttl: Int,
    val protocol: Int,
    val headerChecksum: Int,
    val sourceAddress: ByteArray,
    val destAddress: ByteArray
) {
    val headerLength: Int
        get() = ihl * 4

    val sourceAddressBytes: ByteArray
        get() = sourceAddress.copyOf()

    val destAddressBytes: ByteArray
        get() = destAddress.copyOf()

    /**
     * Serialize the header back to bytes, recalculating the checksum.
     */
    fun toBytes(): ByteArray {
        val bytes = ByteArray(headerLength)

        // Byte 0: version (4 bits) + IHL (4 bits)
        bytes[0] = ((version shl 4) or ihl).toByte()

        // Byte 1: TOS
        bytes[1] = tos.toByte()

        // Bytes 2-3: Total length
        bytes[2] = (totalLength shr 8).toByte()
        bytes[3] = totalLength.toByte()

        // Bytes 4-5: Identification
        bytes[4] = (identification shr 8).toByte()
        bytes[5] = identification.toByte()

        // Bytes 6-7: Flags (3 bits) + Fragment offset (13 bits)
        val flagsAndOffset = (flags shl 13) or fragmentOffset
        bytes[6] = (flagsAndOffset shr 8).toByte()
        bytes[7] = flagsAndOffset.toByte()

        // Byte 8: TTL
        bytes[8] = ttl.toByte()

        // Byte 9: Protocol
        bytes[9] = protocol.toByte()

        // Bytes 10-11: Header checksum (set to 0 for calculation)
        bytes[10] = 0
        bytes[11] = 0

        // Bytes 12-15: Source address
        System.arraycopy(sourceAddress, 0, bytes, 12, 4)

        // Bytes 16-19: Destination address
        System.arraycopy(destAddress, 0, bytes, 16, 4)

        // Calculate and set checksum
        val checksum = Checksum.ipChecksum(bytes, 0, headerLength)
        bytes[10] = (checksum shr 8).toByte()
        bytes[11] = checksum.toByte()

        return bytes
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is IPv4Header) return false
        return version == other.version &&
            ihl == other.ihl &&
            tos == other.tos &&
            totalLength == other.totalLength &&
            identification == other.identification &&
            flags == other.flags &&
            fragmentOffset == other.fragmentOffset &&
            ttl == other.ttl &&
            protocol == other.protocol &&
            headerChecksum == other.headerChecksum &&
            sourceAddress.contentEquals(other.sourceAddress) &&
            destAddress.contentEquals(other.destAddress)
    }

    override fun hashCode(): Int {
        var result = version
        result = 31 * result + ihl
        result = 31 * result + tos
        result = 31 * result + totalLength
        result = 31 * result + identification
        result = 31 * result + flags
        result = 31 * result + fragmentOffset
        result = 31 * result + ttl
        result = 31 * result + protocol
        result = 31 * result + headerChecksum
        result = 31 * result + sourceAddress.contentHashCode()
        result = 31 * result + destAddress.contentHashCode()
        return result
    }

    companion object {
        const val PROTOCOL_TCP = 6
        const val PROTOCOL_UDP = 17

        /**
         * Parse an IPv4 header from raw bytes.
         *
         * @param buffer the byte array containing the IP packet
         * @param offset the starting position within the buffer
         * @return a parsed IPv4Header
         * @throws IllegalArgumentException if the buffer is too short or the version is not 4
         */
        fun parse(buffer: ByteArray, offset: Int = 0): IPv4Header {
            require(buffer.size - offset >= 20) { "Buffer too short for IPv4 header: need 20 bytes, have ${buffer.size - offset}" }

            val versionIhl = buffer[offset].toInt() and 0xFF
            val version = versionIhl ushr 4
            val ihl = versionIhl and 0x0F

            require(version == 4) { "Not an IPv4 packet: version=$version" }
            require(ihl >= 5) { "Invalid IHL: $ihl (minimum is 5)" }
            require(buffer.size - offset >= ihl * 4) { "Buffer too short for IPv4 header with IHL=$ihl" }

            val tos = buffer[offset + 1].toInt() and 0xFF

            val totalLength = ((buffer[offset + 2].toInt() and 0xFF) shl 8) or
                (buffer[offset + 3].toInt() and 0xFF)

            val identification = ((buffer[offset + 4].toInt() and 0xFF) shl 8) or
                (buffer[offset + 5].toInt() and 0xFF)

            val flagsAndOffset = ((buffer[offset + 6].toInt() and 0xFF) shl 8) or
                (buffer[offset + 7].toInt() and 0xFF)
            val flags = flagsAndOffset ushr 13
            val fragmentOffset = flagsAndOffset and 0x1FFF

            val ttl = buffer[offset + 8].toInt() and 0xFF
            val protocol = buffer[offset + 9].toInt() and 0xFF

            val headerChecksum = ((buffer[offset + 10].toInt() and 0xFF) shl 8) or
                (buffer[offset + 11].toInt() and 0xFF)

            val sourceAddress = ByteArray(4)
            System.arraycopy(buffer, offset + 12, sourceAddress, 0, 4)

            val destAddress = ByteArray(4)
            System.arraycopy(buffer, offset + 16, destAddress, 0, 4)

            return IPv4Header(
                version = version,
                ihl = ihl,
                tos = tos,
                totalLength = totalLength,
                identification = identification,
                flags = flags,
                fragmentOffset = fragmentOffset,
                ttl = ttl,
                protocol = protocol,
                headerChecksum = headerChecksum,
                sourceAddress = sourceAddress,
                destAddress = destAddress
            )
        }
    }
}

File: ios/TorVMTunnel/Packet/IPv4Header.swift
import Foundation

/// Shared error type for packet parsing failures.
enum PacketError: Error, CustomStringConvertible {
    case bufferTooShort(needed: Int, have: Int)
    case invalidVersion(UInt8)
    case invalidIHL(UInt8)
    case invalidDataOffset(UInt8)

    var description: String {
        switch self {
        case .bufferTooShort(let needed, let have):
            return "Buffer too short: need \(needed) bytes, have \(have)"
        case .invalidVersion(let v):
            return "Not an IPv4 packet: version=\(v)"
        case .invalidIHL(let ihl):
            return "Invalid IHL: \(ihl) (minimum is 5)"
        case .invalidDataOffset(let offset):
            return "Invalid TCP data offset: \(offset) (minimum is 5)"
        }
    }
}

/// Parses and builds IPv4 headers (minimum 20 bytes).
///
/// All multi-byte fields use network byte order (big-endian).
/// IP addresses are stored as Data(4).
struct IPv4Header {
    static let protocolTCP: UInt8 = 6
    static let protocolUDP: UInt8 = 17

    let version: UInt8
    let ihl: UInt8
    let tos: UInt8
    let totalLength: UInt16
    let identification: UInt16
    let flags: UInt8
    let fragmentOffset: UInt16
    let ttl: UInt8
    let `protocol`: UInt8
    let headerChecksum: UInt16
    let sourceAddress: Data
    let destAddress: Data

    var headerLength: Int {
        return Int(ihl) * 4
    }

    var sourceAddressBytes: Data {
        return Data(sourceAddress)
    }

    var destAddressBytes: Data {
        return Data(destAddress)
    }

    /// Serialize the header back to bytes, recalculating the checksum.
    func toBytes() -> Data {
        var bytes = Data(count: headerLength)

        // Byte 0: version (4 bits) + IHL (4 bits)
        bytes[0] = (version << 4) | ihl

        // Byte 1: TOS
        bytes[1] = tos

        // Bytes 2-3: Total length
        bytes[2] = UInt8(totalLength >> 8)
        bytes[3] = UInt8(totalLength & 0xFF)

        // Bytes 4-5: Identification
        bytes[4] = UInt8(identification >> 8)
        bytes[5] = UInt8(identification & 0xFF)

        // Bytes 6-7: Flags (3 bits) + Fragment offset (13 bits)
        let flagsAndOffset = (UInt16(flags) << 13) | fragmentOffset
        bytes[6] = UInt8(flagsAndOffset >> 8)
        bytes[7] = UInt8(flagsAndOffset & 0xFF)

        // Byte 8: TTL
        bytes[8] = ttl

        // Byte 9: Protocol
        bytes[9] = self.`protocol`

        // Bytes 10-11: Header checksum (set to 0 for calculation)
        bytes[10] = 0
        bytes[11] = 0

        // Bytes 12-15: Source address
        bytes.replaceSubrange(12..<16, with: sourceAddress)

        // Bytes 16-19: Destination address
        bytes.replaceSubrange(16..<20, with: destAddress)

        // Calculate and set checksum
        let checksum = Checksum.ipChecksum(data: bytes, offset: 0, length: headerLength)
        bytes[10] = UInt8(checksum >> 8)
        bytes[11] = UInt8(checksum & 0xFF)

        return bytes
    }

    /// Parse an IPv4 header from raw bytes.
    ///
    /// - Parameters:
    ///   - data: the data containing the IP packet
    ///   - offset: the starting position within the data
    /// - Returns: a parsed IPv4Header
    /// - Throws: PacketError if the buffer is too short or the version is not 4
    static func parse(data: Data, offset: Int = 0) throws -> IPv4Header {
        guard data.count - offset >= 20 else {
            throw PacketError.bufferTooShort(needed: 20, have: data.count - offset)
        }

        let versionIhl = data[offset]
        let version = versionIhl >> 4
        let ihl = versionIhl & 0x0F

        guard version == 4 else {
            throw PacketError.invalidVersion(version)
        }
        guard ihl >= 5 else {
            throw PacketError.invalidIHL(ihl)
        }
        guard data.count - offset >= Int(ihl) * 4 else {
            throw PacketError.bufferTooShort(needed: Int(ihl) * 4, have: data.count - offset)
        }

        let tos = data[offset + 1]

        let totalLength = (UInt16(data[offset + 2]) << 8) |
            UInt16(data[offset + 3])

        let identification = (UInt16(data[offset + 4]) << 8) |
            UInt16(data[offset + 5])

        let flagsAndOffset = (UInt16(data[offset + 6]) << 8) |
            UInt16(data[offset + 7])
        let flags = UInt8(flagsAndOffset >> 13)
        let fragmentOffset = flagsAndOffset & 0x1FFF

        let ttl = data[offset + 8]
        let proto = data[offset + 9]

        let headerChecksum = (UInt16(data[offset + 10]) << 8) |
            UInt16(data[offset + 11])

        let sourceAddress = data.subdata(in: (offset + 12)..<(offset + 16))
        let destAddress = data.subdata(in: (offset + 16)..<(offset + 20))

        return IPv4Header(
            version: version,
            ihl: ihl,
            tos: tos,
            totalLength: totalLength,
            identification: identification,
            flags: flags,
            fragmentOffset: fragmentOffset,
            ttl: ttl,
            protocol: proto,
            headerChecksum: headerChecksum,
            sourceAddress: sourceAddress,
            destAddress: destAddress
        )
    }
}
